core/tx_pool.go

```diff
@@ -658,15 +658,23 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	if pool.currentState.GetNonce(from) > tx.Nonce() {
 		return ErrNonceTooLow
 	}
-	// Get L1 data fee in current state
-	l1DataFee, err := fees.CalculateL1DataFee(tx, pool.currentState)
-	if err != nil {
-		return fmt.Errorf("failed to calculate L1 data fee, err: %w", err)
-	}
-	// Transactor should have enough funds to cover the costs
-	// cost == V + GP * GL
-	if pool.currentState.GetBalance(from).Cmp(new(big.Int).Add(tx.Cost(), l1DataFee)) < 0 {
-		return ErrInsufficientFunds
+	if pool.chainconfig.Scroll.FeeVaultEnabled() {
+		// Get L1 data fee in current state
+		l1DataFee, err := fees.CalculateL1DataFee(tx, pool.currentState)
+		if err != nil {
+			return fmt.Errorf("failed to calculate L1 data fee, err: %w", err)
+		}
+		// Transactor should have enough funds to cover the costs
+		// cost == V + GP * GL + L1 data fee
+		if b := pool.currentState.GetBalance(from); b.Cmp(new(big.Int).Add(tx.Cost(), l1DataFee)) < 0 {
+			return ErrInsufficientFunds
```

Comment:
Strictly speaking this is a breaking API change, since it changes the returned error. Can we preserve the original behaviour? (Return `invalid transaction: insufficient funds for l1fee + gas * price + value`.)

Reply:
nice catch, fixed in [fc18a07](https://github.com/scroll-tech/go-ethereum/pull/612/commits/fc18a0756cc80f683ba06a8228fde4343f612b8a)